### PR TITLE
Decouple Loader from Url Format and Move Parsing to Resolver

### PR DESCRIPTION
**Pull request by anthony-murphy**
There are implicit dependencies on the resolved url format which forced thing like `placeholder/placeholder/` in the resolved odsp url. This change makes id a first class property on the resolve url, and reduces our dependency on string parsing, and implicit contracts.

### CODE REVIEW

File: packages/loader/container-loader/src/container.ts [hunk 1]
```diff
@@ -128,13 +126,13 @@ export interface IContainerLoadOptions {
...
 export interface IContainerConfig {
-    resolvedUrl?: IResolvedUrl;
+    resolvedUrl?: IFluidResolvedUrl;
     canReconnect?: boolean;
     /**
      * A url for the Container.  Critically, we expect Loader.resolve using this URL to resolve back to this Container
      * for purposes of creating a separate Container instance for the summarizer to use.
      */
-    containerUrl?: string;
+         containerUrl?: string;
```
Comment (ChumpChief) [inline packages/loader/container-loader/src/container.ts:135]:
```suggestion
    containerUrl?: string;
```

File: packages/loader/container-loader/src/loader.ts [hunk 2]
```diff
@@ -305,14 +305,16 @@ export class Loader extends EventEmitter implements IHostLoader {
...
             const resolved = await this.resolveCore(request);
-            return resolved.container.request({ url: `${resolved.parsed.path}${resolved.parsed.query}` });
+            const parsed = parseUrl(resolved.resolved.url);
+            assert(parsed !== undefined, "resolve url could not be parsed");
+            return resolved.container.request({ url: `${parsed.path}${parsed.query}` });
         });
     }
 
-    private getKeyForContainerCache(request: IRequest, parsedUrl: IParsedUrl): string {
+    private getKeyForContainerCache(request: IRequest, resolverUrl: IFluidResolvedUrl): string {
```
Comment (ChumpChief) [inline packages/loader/container-loader/src/loader.ts:314]:
```suggestion
    private getKeyForContainerCache(request: IRequest, resolvedUrl: IFluidResolvedUrl): string {
```

**Review by ChumpChief — APPROVED** [review]
Very nice :)

**Comment by anthony-murphy** [comment]
#5794 is the next step here, but this unblocks some of the work to simplify the relative loading flow

**Review by jatgarg — APPROVED** [review]

File: packages/drivers/odsp-driver/src/test/odspDriverResolverTest.spec.ts [hunk 3]
```diff
@@ -182,8 +186,8 @@ describe("Odsp Driver Resolver", () => {
         assert.strictEqual(resolvedUrl.hashedDocumentId, "", "No doc id should be present");
         assert.strictEqual(resolvedUrl.endpoints.snapshotStorageUrl, "", "Snapshot url should be empty");
 
-        const expectedResolvedUrl = `fluid-odsp://${siteUrl}?driveId=${driveId}&path=${testFilePath}&itemId=${itemId}`
-        + `&version=null`;
+        const expectedResolvedUrl = `fluid-odsp://localhost?driveId=${driveId}&path=${testFilePath}&itemId=${itemId}`
+        + ``;
```
Comment (anthony-murphy) [inline packages/drivers/odsp-driver/src/test/odspDriverResolverTest.spec.ts:190]:
```suggestion
```

File: packages/drivers/odsp-driver/src/test/odspDriverResolverTest.spec.ts [hunk 4]
```diff
@@ -182,8 +186,7 @@ describe("Odsp Driver Resolver", () => {
         assert.strictEqual(resolvedUrl.hashedDocumentId, "", "No doc id should be present");
         assert.strictEqual(resolvedUrl.endpoints.snapshotStorageUrl, "", "Snapshot url should be empty");
 
-        const expectedResolvedUrl = `fluid-odsp://${siteUrl}?driveId=${driveId}&path=${testFilePath}&itemId=${itemId}`
-        + `&version=null`;
+        const expectedResolvedUrl = `fluid-odsp://localhost?driveId=${driveId}&path=${testFilePath}&itemId=${itemId}`
```
Comment (anthony-murphy) [inline packages/drivers/odsp-driver/src/test/odspDriverResolverTest.spec.ts:189]:
```suggestion
        const expectedResolvedUrl = `fluid-odsp://localhost?driveId=${driveId}&path=${testFilePath}&itemId=${itemId}`;
```

File: packages/drivers/odsp-driver/src/test/odspCreateContainer.spec.ts [hunk 5]
```diff
@@ -90,7 +90,7 @@ describe("Odsp Create Container Test", () => {
         assert.strictEqual(finalResolverUrl.siteUrl, siteUrl, "SiteUrl should match");
         assert.strictEqual(finalResolverUrl.hashedDocumentId, docID, "DocId should match");
 
-        const url = `fluid-odsp://placeholder/placeholder/${
+        const url = `fluid-odsp://${
```
Comment (anthony-murphy) [inline packages/drivers/odsp-driver/src/test/odspCreateContainer.spec.ts:93]:
```suggestion
        const url = `fluid-odsp://placeholder/placeholder/${
```

File: packages/test/test-end-to-end-tests/src/test/pendingOps.spec.ts [hunk 6]
```diff
@@ -317,7 +317,7 @@ describeNoCompat("stashed ops", (getTestObjectProvider) => {
             const router = await runtime.createDataStore(["default"]);
             const dataStore = await requestFluidObject<ITestFluidObject>(router, "/");
 
-            const channel = dataStore.runtime.createChannel(newMapId, "https://graph.microsoft.com/types/map");
+        const channel = dataStore.runtime.createChannel(newMapId, "https://graph.microsoft.com/types/map");
```
Comment (anthony-murphy) [inline packages/test/test-end-to-end-tests/src/test/pendingOps.spec.ts:320]:
```suggestion
            const channel = dataStore.runtime.createChannel(newMapId, "https://graph.microsoft.com/types/map");
```

**Pull request merged**